src/YetAnotherHttpHandler/package.json

```diff
@@ -5,5 +5,9 @@
   "displayName": "YetAnotherHttpHandler",
   "description": "YetAnotherHttpHandler that brings the power of HTTP/2 (and gRPC) to Unity and .NET Standard.",
   "unity": "2021.3",
-  "license": "MIT"
+  "license": "MIT",
+  "dependencies": {
+    "org.nuget.system.io.pipelines": "7.0.0",
+    "org.nuget.system.runtime.compilerservices.unsafe": "4.5.3"
+  }
```

Comment:
If install the package by referring to git without using a scoped registry, it seems to cause an error due to these dependencies. 
```
[Package Manager Window] Cannot perform upm operation: Unable to add package [https://github.com/bdovaz/YetAnotherHttpHandler.git?path=src/YetAnotherHttpHandler#feature/resolve-dependencies-using-unity-nuget]:
  Package com.cysharp.yetanotherhttphandler@https://github.com/bdovaz/YetAnotherHttpHandler.git?path=src/YetAnotherHttpHandler#feature/resolve-dependencies-using-unity-nuget has invalid dependencies or related test packages:
    org.nuget.system.io.pipelines (dependency): Package [org.nuget.system.io.pipelines@7.0.0] cannot be found [NotFound].
UnityEditor.EditorApplication:Internal_CallUpdateFunctions ()
```

Need to create a built-in package for dependency reference and change it to be referred from YetAnotherHttpHandler.Unity project.

Reply:
@mayuki but of course, this problem occurs if you have not imported before the *.unitypackage which is precisely the one that includes those packages so that then the error does not appear.

If you try to directly install the package without installing the dependencies this will happen.

Precisely for that reason in this PR apart from adding a second installation method (UnityNuGet) I have also inverted the order of the steps, that is to say, in the case of the installation without UnityNuGet I have expressly put that first you have to import the *.unitypackage file of the dependencies.

Reply:
Thank you. I understand the reason for the problem.

Our policy is to avoid creating dependencies on other packages.
This is because it becomes a problem if the dependency is already managed by another mechanism, such as internal UPM.

In our development environment, libraries are distributed under their own internal package names. For example, if `System.Buffers` has already been distributed/installed (e.g. `com.example.system.buffers` package), other dependent libraries (Grpc.Net.Client, System.Memory ...) can be moved from `Packages` to `Plugins` to ignore package dependencies, but YetAnotherHttpHandler will fail to resolve dependencies.

Reply:
But what I don't understand is, this problem I didn't create it already existed.

I mean...

Actually, in the `main` branch it says to first import the package with Git and then import the unitypackage with the dependencies. Even if it allows you to install the Git package because the package.json has no dependencies, it will not compile because you have not imported the unitypackage file before.

That is, the order of the instructions should be reversed as I have done in this PR in the Readme.md.

And on the other hand, if a package really has dependencies why shouldn't they be put? I don't understand, that's how this UPM / NPM system works, it's the natural thing to do.

To work with OpenUPM / UnityNuGet it is necessary that the versioned package.json file has them, otherwise this PR loses all sense and all the time invested by me.

I'm really trying hard to make it work "as before" and through this OpenUPM / UnityNuGet flow that has become almost a standard among developers.

What do you suggest to keep this possible?

Reply:
Ah, I see. My apologies for misunderstanding.

Given the use of UnityNuGet, it's unavoidable to list dependencies here.
As a result, I understand that if you install via git and without using UnityNuGet, it's necessary to install a .unitypackage that includes the `org.nuget.*` packages.

On the other hand, the problem I was having is when a dependency is already installed from elsewhere and the package cannot be installed by UnityNuGet or .unitypackage. Ignore the package when Unity Package Manager cannot resolve the dependency, I wish I could...

While I believe that directing to install from a .unitypackage will solve the problem in most use cases, could there be a solution to address the following case?

```mermaid
flowchart TB
    UnityProject-->YetAnotherHttpHandler
    UnityProject-->InternalLib

    YetAnotherHttpHandler --"org.nuget.system.buffers"--> System.IO.Pipelines
    YetAnotherHttpHandler --"org.nuget.system.runtime.compilerservices.unsafe"--> System.Runtime.CompilerServices.Unsafe

    subgraph InternalUPM
        InternalLib--"com.example.system.buffers"-->System.Buffers
    end
    subgraph .unitypackage
        System.IO.Pipelines --X--> System.Buffers
        System.IO.Pipelines --"org.nuget.system.memory"--> System.Memory
        System.IO.Pipelines --"org.nuget.system.threading.tasks.extensions"--> System.Threading.Tasks.Exensions
 
        System.Memory --X--> System.Buffers
        System.Memory --"org.nuget.system.numerics.vectors"--> System.Numerics.Vectors
        System.Memory --"org.nuget.system.runtime.compilerservices.unsafe"--> System.Runtime.CompilerServices.Unsafe
    end
```

Reply:
Of course, it would be best to either remove the dependency from the internal library or change it to distribute with the same package ID. 

The workarounds I can think of are to add a dummy package on the game project side, or prepare two different package.json on YetAnotherHttpHandler side.

Reply:
For the more specific cases you say that can not / want to install it through any of the 2 options (UnityNuGet or unitypackage) and want to install them by hand, there is always a solution.

And it is, embed the package: https://docs.unity3d.com/Manual/upm-embed.html

The most common case is when you want to fix a bug in the package or you have a collision between assemblies.

We can document this use case in the `Readme.md` but we should not do more than that, they are much more niche cases that we should not complicate the solution for it.

Reply:
@bdovaz 
I apologize for the delayed response!

Our internal feedback is that it is better to have no dependencies at the time of distribution.
This is due to the occurrence of conflicts with internal packages.
I agree that it is correct for dependencies to be described in the package manifest, and it's unnatural if they're not.

However, on the other hand, the current Unity package ecosystem is very incomplete, and all of the package registries other than the official Unity ones are by third parties.
Unlike npm and NuGet, there are various package registries, both internal and external, and conflicts often occur.

Considering these circumstances, we had a policy of not depend on specific package registries such as OpenUPM or UnityNuGet, including publishing them, in order to ensure flexibility for our users when providing Unity libraries.
We still want to maintain this policy, so we would like to specify no dependencies again this time.

On the other hand, I understand that the ease of installation using UnityNuGet is a benefit for users, and that creating a `.unitypackage` is also easier. I also do not want this PR you have contributed to end up being wasted.

Therefore, what about having the `YetAnotherHttpHandler/package.json` without dependencies, and instead making `YetAnotherHttpHandler.Unity.Dependencies` a Unity package and specify the dependencies there?

You might not agree with needing to install `YetAnotherHttpHandler.Unity.Dependencies` in addition to `YetAnotherHttpHandler` when using UnityNuGet, but I believe that this way, your PR will not be waste.

Reply:
The second package `YetAnotherHttpHandler.Unity.Dependencies` is a package solely for dependency resolution for users using UnityNuGet. It is not intended for use with OpenUPM.

The aim is to complete the installation of the package from the two URLs below after configuring UnityNuGet.

```
https://github.com/Cysharp/YetAnotherHttpHandler.git?path=src/YetAnotherHttpHandler
https://github.com/Cysharp/YetAnotherHttpHandler.git?path=src/YetAnotherHttpHandler.Dependencies
```

This proposal aims to keep an easy path for UnityNuGet users while there is a unique circumstance of not wanting to specifiy dependencies on `YetAnotherHttpHandler`.

Reply:
@mayuki okay, and do you have an idea where to create it? Because that folder name you already have it reserved and it has a csproj in it.

https://github.com/Cysharp/YetAnotherHttpHandler/tree/main/src/YetAnotherHttpHandler.Unity.Dependencies

Reply:
The current YetAnotherHttpHandler.Unity.Dependencies is a development helper tool project for copying DLLs.
I think it's okay to replace its contents because its role is ended by using UnityNuGet.

Reply:
@mayuki seeing the `YetAnotherHttpHandler.Unity.Dependencies.Sync` project loses its sense no? Because it is already going to be solved from UnityNuGet.

Reply:
Yes, that's correct, it will become completely unnecessary.